lib/internal/errors.js

```diff
@@ -622,6 +622,45 @@ function addNumericalSeparator(val) {
   return `${val.slice(0, i)}${res}`;
 }
 
+// Used to enhance the stack that will be picked up by the inspector
+const kEnhanceStackBeforeInspector = Symbol('kEnhanceStackBeforeInspector');
+
+// These are supposed to be called only on fatal exceptions before
+// the process exits.
+const fatalExceptionStackEnhancers = {
+  beforeInspector(error) {
+    if (typeof error[kEnhanceStackBeforeInspector] !== 'function') {
+      return error.stack;
+    }
+
+    try {
+      // Set the error.stack here so it gets picked up by the
+      // inspector.
+      error.stack = error[kEnhanceStackBeforeInspector]();
+    } catch {
+      // We are just enhancing the error. If it fails, ignore it.
+    }
+    return error.stack;
+  },
+  afterInspector(error) {
+    const originalStack = error.stack;
+    const {
+      inspect,
+      inspectDefaultOptions: {
+        colors: defaultColors
+      }
+    } = lazyInternalUtilInspect();
+    const colors = internalBinding('util').guessHandleType(2) === 'TTY' &&
+                   require('internal/tty').hasColors() ||
+                   defaultColors;
+    try {
+      return inspect(error, { colors });
+    } catch {
+      return originalStack;
```

Comment:
Note that here we don't patch `error.stack` at all - patching `error.stack` is only one of the many ways to pass some metadata around along with the error. The cleanest way, however, is to just wrap those metadata in something else instead of polluting `error` itself over and over, which is also what V8 does in its embedder API (the source of truth of the stack trace is `message->GetStackTrace()`, not `error->Get(context, stack_string)`.

Reply:
Can't we do this in a WeakMap? Their point is extending things from the outside after all

Reply:
I don't think we need a WeakMap for this, as this is kind of a one-off thing so there is no mapping necessary. We only need to pass a separate object around, the association between the error and the object can already be known from how they are passed around, and we probably don't need an equivalent of `Exception::CreateMessage`.

Reply:
(Also now that you mentioned it...I believe V8 still does not have embedder API to interface with  WeakMaps? That's why we still have to create our ESM module wrap map in JS land)